### DISCUSSION TITLE
Testing maximumSize?

### DISCUSSION THREAD

**Discussion by Blackbaud-JasonBodnar** [ben-manes/caffeine#717] [Q&A] [answered]
I've created a cache with a max size of 500:

```
CaffeineCache reCountriesCache = new CaffeineCache(
    "reCountries",
    Caffeine.newBuilder()
        .ticker(ticker)
        .expireAfterWrite(Duration.of(3, ChronoUnit.DAYS))
        .expireAfterAccess(Duration.of(1, ChronoUnit.DAYS))
        .maximumSize(500)
        .scheduler(Scheduler.systemScheduler())
        .build()
);
```

```
@Cacheable(cacheNames = "reCountries")
public List<RECountry> getReCountries(String environmentId) {
    return RequestContext.withUserAndEnvironment(SYSTEM_USER_ID, environmentId, () -> tributeAddressClient.getCountryList()).getValue();
}
```

I'm trying to test that the max size is working:

```
given:
RECountryCollection cachedCollection = aRandom.reCountryCollection().build()

when:
reCountryService.getReCountries(environmentId)

then:
1 * mockTributeAddressClient.getCountryList() >> cachedCollection

when:
reCountryService.getReCountries(environmentId)

then:
0 * mockTributeAddressClient.getCountryList()

when:
501.times { reCountryService.getReCountries(aRandom.environmentId()) }

then:
501 * mockTributeAddressClient.getCountryList() >> aRandom.reCountryCollection().build()

when:
reCountryService.getReCountries(environmentId)

then:
1 * mockTributeAddressClient.getCountryList() >> cachedCollection
```

When I run the test the cache is being used in the last when but I was expecting the actual method to be called since the cache should now have hit the 500 entry limit.

**Comment by ben-manes** [comment, marked as answer]
I am not really following the code to be certain what the question is. I think you are observing more than 500 entries cached, which is allowed temporarily by `Caffeine.maximumSize` javadoc to better cope with concurrency.

The cache defers eviction to the background executor, which by default is `ForkJoinPool.commonPool()`. In your case the test thread is likely running before the maintenance kicks in. If you use `Caffeine.executor(Runnable::run)` then it will run on the caller. The cache's own operations are inexpensive, but due to user callbacks we can't know those penalties. Since usually one does not inspect the cache this is a good default, but can be annoying for tests.

The [Testing](https://github.com/ben-manes/caffeine/wiki/Testing) wiki page might be of interest. Please let me know if I misunderstood and did not answer your question.

**Comment by Blackbaud-JasonBodnar** [comment]
I'm trying to test the 500 capacity by seeding the cache with one entry (the first `when`/`then` combination), asserting the entry is in the cache (the second `when`/`then` combo), adding 501 distinct entries to the cache (the 3rd `when`/`then` combo) and finally proving that the first entry is no longer cached by getting it again and expecting the cache not to be used (the last `when`/`then` combo).

I guess I'm thinking too simplistically due to what you mentioned with things running in separate threads and also possibly that there's no guarantee that the first entry would be the one to be evicted.

I personally don't like doing test like this as it's really testing the library which we know is already thoroughly tested, but our team wants to test the cache is configured correctly. Is there a way to verify the various settings of a cache? We are using Spring's caching so I don't necessarily have direct access to the caffeine cache but rather Spring's `CacheManager`.

**Reply by ben-manes** [reply]
You're right that makes assumptions about the eviction policy. Since we a more advanced one than LRU, it is likely to notice that the new arrivals are one-hit wonders and prefer to retain the first entries (e.g. scans are common in databases where an MRU would be ideal policy).

I haven't used Spring in a long time, so unfortunately I am not positive. For a unit test you could provide a concrete `CaffeineCacheManager` with your settings. Or you might be able to hook into their cache statistics to verify an eviction took place (assuming some handling for the above mentioned async default).

**Comment by Blackbaud-JasonBodnar** [comment]
For anybody looking for help testing a caffeine cache configuration while using Spring's `CacheManager` here's a solution (in groovy):

```
Cache cache = cacheManager.getCache("reCountries").nativeCache as Cache
assert cache.policy().eviction().get().maximum == 500
assert cache.policy().expireAfterWrite().get().expiresAfter == Duration.of(3, ChronoUnit.DAYS)
```
